### PR TITLE
Refactor water configuration json files and configuration readers

### PR DESCRIPTION
**Pull request by snb3300**
# Description:

Logic to read config file in ```ComplexWaterEntityFinder``` was tightly coupled with only one configuration file format. Due to this restriction, creating ```ComplexWaterEntity``` was not flexible. This PR add a way to specify a custom configuration file and the associated reader for that file. This makes it easy to implement custom configuration readers and pass them to ```ComplexWaterEntityFinder```. 

This PR also updates ```ComplexIslandFinder``` so that it is not dependent on the water layer and it can accept a custom configuration file and the corresponding configuration reader.

### Potential Impact:

No potential impact on atlas-checks or atlas generator.

### Unit Test Approach:

1. Existing unit tests cover a lot of test cases
2. New unit tests added to separate out island from the water

### Test Results:

All unit tests passed

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/water/finder/ComplexWaterEntityFinder.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    public ComplexWaterEntityFinder(final WaterConfigurationReader... waterConfigurationReaders)
+    {
+        // use the passed in configuration readers for the filters
+        this.waterConfigurationReaders = Arrays.asList(waterConfigurationReaders);
+    }
+
+    @Override
+    public Iterable<ComplexWaterEntity> find(final Atlas atlas)
+    {
+
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/water/finder/ComplexWaterEntityFinder.java]: Whitespace

File: src/main/resources/org/openstreetmap/atlas/geography/atlas/items/complex/islands/islands.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (matthieun) [inline src/main/resources/org/openstreetmap/atlas/geography/atlas/items/complex/islands/islands.json]: Any reason this one is in a different package than the others?

File: src/main/resources/org/openstreetmap/atlas/geography/atlas/items/complex/islands/islands.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  "global": {
+    "scanUrls": [
```
Comment (matthieun) [inline src/main/resources/org/openstreetmap/atlas/geography/atlas/items/complex/islands/islands.json]: Is that necessary if there is an "import" statement below?

File: src/main/resources/org/openstreetmap/atlas/geography/atlas/items/complex/islands/islands.json (hunk 2 above)
Comment (snb3300) [reply src/main/resources/org/openstreetmap/atlas/geography/atlas/items/complex/islands/islands.json]: I tried to keep islands and water separate, hence different packages.

**Comment by snb3300** [comment]
I think they were there before. I'll fix them @matthieun

**Comment by matthieun** [comment]
Aah sonar. You implement equals and hashcode and then it complains of duplication..

**Pull request merged**